ext/nnstreamer/extra/nnstreamer_protobuf_grpc.cc

```diff
@@ -0,0 +1,371 @@
+    void setConfig (GstTensorsConfig *config) {
+      memcpy (&config_, config, sizeof (*config));
+    }
+
+  private:
+    /** @brief start server service */
+    gboolean _start_server () {
+      std::string address (host_);
+
+      address += ":" + std::to_string (port_);
+
+      grpc::EnableDefaultHealthCheckService(true);
+
+      /* listen on the given address without any authentication mechanism */
+      ServerBuilder builder;
+      builder.AddListeningPort (address, grpc::InsecureServerCredentials());
+      builder.RegisterService (this);
+
+      /* start the server */
+      server_ = builder.BuildAndStart ();
+      if (server_.get () == nullptr)
+        return FALSE;
+
+      return TRUE;
+    }
+
+    /** @brief start client service */
+    gboolean _start_client () {
+      std::string address (host_);
+
+      address += ":" + std::to_string (port_);
+
+      /* create a gRPC channel */
+      std::shared_ptr<Channel> channel = grpc::CreateChannel(
+          address, grpc::InsecureChannelCredentials());
+
+      /* connect the server */
+      stub_ = TensorService::NewStub (channel);
+      if (stub_.get () == nullptr)
+        return FALSE;
+
+      queue_ = gst_data_queue_new (_data_queue_check_full_cb, NULL, NULL, NULL);
+      worker_ = std::thread ([this] { this->_writer_thread (); });
+
+      return TRUE;
+    }
+
+    /** @brief convert buffer to tensors */
+    void _get_tensors_from_buffer (GstBuffer *buffer, Tensors &tensors) {
+      Tensors::frame_rate *fr;
+      GstMapInfo map;
+      gsize data_ptr = 0;
+
+      tensors.set_num_tensor (config_.info.num_tensors);
+
+      fr = tensors.mutable_fr ();
+      fr->set_rate_n (config_.rate_n);
+      fr->set_rate_d (config_.rate_d);
+
+      gst_buffer_map (buffer, &map, GST_MAP_READ);
+
+      for (guint i = 0; i < config_.info.num_tensors; i++) {
+        nnstreamer::protobuf::Tensor *tensor = tensors.add_tensor ();
+        const GstTensorInfo * info = &config_.info.info[i];
+        gsize tsize = gst_tensor_info_get_size (info);
+
+        /* set tensor info */
+        tensor->set_name ("Anonymous");
+        tensor->set_type ((Tensor::Tensor_type) info->type);
+
+        for (guint j = 0; j < NNS_TENSOR_RANK_LIMIT; j++)
+          tensor->add_dimension (info->dimension[j]);
+
+        if (data_ptr + tsize > map.size) {
+          ml_logw ("Setting invalid tensor data");
+          break;
+        }
+
+        tensor->set_data (map.data + data_ptr, tsize);
+        data_ptr += tsize;
+      }
+
+      gst_buffer_unmap (buffer, &map);
+    }
+
+    /** @brief gRPC writer thread */
+    void _writer_thread () {
+      ClientContext context;
+      google::protobuf::Empty reply;
+
+      /* initiate the RPC call */
+      std::unique_ptr< ClientWriter<Tensors> > writer(
+          stub_->SendTensors (&context, &reply));
+
+      GstDataQueueItem *item;
+      GstBuffer *buffer;
+
+      /* until flushing */
+      while (gst_data_queue_pop (queue_, &item)) {
```

Comment:
Q: Is it possible to block forever?

Reply:
Well, it's unblocked before `stop` is called. I'm not sure that it's always guaranteed in gstreamer internals.